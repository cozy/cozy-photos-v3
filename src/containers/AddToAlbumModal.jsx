import styles from '../styles/addToAlbum'

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { translate } from '../lib/I18n'
import Modal, { ModalSection } from 'cozy-ui/react/Modal'
import classNames from 'classnames'

import Alerter from '../components/Alerter'
import CreateAlbumForm from '../components/CreateAlbumForm'
import SelectAlbumsForm from '../components/SelectAlbumsForm'
import Loading from '../components/Loading'

import {
  fetchAlbums,
  getAlbumsList,
  cancelAddToAlbum,
  closeAddToAlbum,
  createAlbum,
  addToAlbum }
from '../ducks/albums'

export class AddToAlbumModal extends Component {
  componentWillMount () {
    this.props.fetchAlbums()
  }

  render (props) {
    const {
      t,
      photos,
      albums,
      onDismiss,
      onSubmitNewAlbum,
      onSubmitSelectedAlbum
    } = props

    const { fetchingStatus } = props.albums ? props.albums : { fetchingStatus: 'loading' }
    const isFetchingAlbums = fetchingStatus === 'pending' || fetchingStatus === 'loading'

    return (
      <Modal
        title={t('Albums.add_photos.title')}
        secondaryAction={() => onDismiss()}
        >
        <ModalSection className={styles['coz-modal-section']}>
          <div className={classNames(styles['coz-create-album'])}>
            <CreateAlbumForm
              onSubmitNewAlbum={name => onSubmitNewAlbum(name, photos)}
              />
          </div>
        </ModalSection>
        {
          isFetchingAlbums
          ? <ModalSection><Loading loadingType='albums_fetching' /></ModalSection>
          : albums && albums.entries.length > 0
          ? <ModalSection><div className={classNames(styles['coz-select-album'])}>
            <SelectAlbumsForm albums={albums} onSubmitSelectedAlbum={album => onSubmitSelectedAlbum(album, photos)} />
          </div></ModalSection>
          : null
        }
      </Modal>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    photos: state.ui.selected,
    albums: getAlbumsList(state)
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDismiss: () => {
    dispatch(cancelAddToAlbum())
  },
  onSubmitNewAlbum: (name, photos) => {
    return dispatch(createAlbum(name, photos))
      .then(() => {
        dispatch(closeAddToAlbum())
        Alerter.success('Albums.create.success', {name: name, smart_count: photos.length})
      })
  },
  onSubmitSelectedAlbum: (album, photos) => {
    return dispatch(addToAlbum(album, photos))
      .then(() => {
        dispatch(closeAddToAlbum())
        Alerter.success('Albums.add_photos.success', {name: album.name, smart_count: photos.length})
      })
  },
  fetchAlbums: () => dispatch(fetchAlbums())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AddToAlbumModal))
