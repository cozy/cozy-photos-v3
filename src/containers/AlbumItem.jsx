/* global cozy */
import styles from '../styles/albumsList'
import classNames from 'classnames'

import React, { Component } from 'react'
import { Link, withRouter } from 'react-router'
import { cozyConnect } from '../lib/redux-cozy-client'

import { fetchAlbumPhotos } from '../ducks/albums'
import ImageLoader from '../components/ImageLoader'

const SharedIcon = ({ byMe }) => (
  <div className={styles['pho-album-shared']}>
    <div className={classNames(styles['pho-album-shared-icon'], styles[byMe ? '--by-me' : '--with-me'])} />
  </div>
)

const AlbumItemLink = ({ router, album, image, title, desc }) => {
  const parentPath = router.location.pathname
  return (
    <Link to={`${parentPath}/${album._id}`} className={styles['pho-album-link']}>
      {image}
      {title}
      {desc}
    </Link>
  )
}
const LinkedAlbumItem = withRouter(AlbumItemLink)

const ClickableAlbumItem = ({ album, image, title, desc, onClick, disabled }) => {
  return disabled
  ? <div className={classNames(styles['pho-album-link'], styles['pho-album-link--disabled'])}>
    {image}
    <div>
      {title}
      {desc}
    </div>
  </div>
  : <div onClick={() => onClick(album)} className={styles['pho-album-link']}>
    {image}
    <div>
      {title}
      {desc}
    </div>
  </div>
}

export class AlbumItem extends Component {
  render () {
    const { t, album, photos, sharedByMe, sharedWithMe, onClick, disabled } = this.props
    if (photos.fetchStatus !== 'loaded') {
      return null
    }

    const coverPhoto = photos.data[0] || null
    const image = !coverPhoto
      ? <div
        className={styles['pho-album-photo-item']}
      />
      : <ImageLoader
        className={styles['pho-album-photo-item']}
        alt={`${album.name} album cover`}
        photo={coverPhoto}
        src={`${cozy.client._url}${coverPhoto.links.small}`}
      />

    const desc = <h4 className={styles['pho-album-description']}>
      {(sharedByMe || sharedWithMe) && <SharedIcon byMe={sharedByMe} />}
      {t('Albums.album_item_description',
        {smart_count: photos.count})
      }
      {(sharedByMe || sharedWithMe) && ` - ${t('Albums.album_item_shared_ro')}`}
    </h4>

    const title = <h2 className={styles['pho-album-title']}>{album.name}</h2>

    return (
      <div className={styles['pho-album']}>
        {
          onClick
          ? <ClickableAlbumItem album={album} image={image} title={title} desc={desc} onClick={onClick} disabled={disabled} />
          : <LinkedAlbumItem album={album} image={image} title={title} desc={desc} />
        }
      </div>
    )
  }
}

const mapDocumentsToProps = (ownProps) => ({
  photos: fetchAlbumPhotos(ownProps.album)
})

export default cozyConnect(mapDocumentsToProps)(AlbumItem)
