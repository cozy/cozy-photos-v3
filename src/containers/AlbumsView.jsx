import React, { Component } from 'react'
import { cozyConnect } from '../lib/redux-cozy-client'
import styles from '../styles/layout'

import { AlbumsToolbar, fetchAlbums } from '../ducks/albums'
import { filterSharedByLinkDocuments, filterSharedWithMeDocuments } from '../ducks/sharing'

import AlbumsList from '../components/AlbumsList'
import Loading from '../components/Loading'
import ErrorComponent from '../components/ErrorComponent'
import Topbar from '../components/Topbar'

const Content = ({ list, sharedByMe, sharedWithMe }) => {
  const { fetchStatus, data } = list
  switch (fetchStatus) {
    case 'pending':
    case 'loading':
      return <Loading loadingType='albums_fetching' />
    case 'failed':
      return <ErrorComponent errorType='albums' />
    default:
      return <AlbumsList albums={data} sharedByMe={sharedByMe} sharedWithMe={sharedWithMe} />
  }
}

export class AlbumsView extends Component {
  constructor (props) {
    super(props)
    this.state = {
      sharedByMe: [],
      sharedWithMe: []
    }
  }

  componentWillReceiveProps (newProps) {
    // TODO: this is not the cleanest way of doing this...
    if (newProps.albums && newProps.albums.data !== 0) {
      Promise.all([
        filterSharedByLinkDocuments(newProps.albums.data.map(a => a.id), 'io.cozy.photos.albums'),
        filterSharedWithMeDocuments(newProps.albums.data.map(a => a.id), 'io.cozy.photos.albums')
      ])
        .then(sharedIds => this.setState({ sharedByMe: sharedIds[0], sharedWithMe: sharedIds[1] }))
    }
  }

  render () {
    if (this.props.children) return this.props.children
    if (!this.props.albums) {
      return null
    }
    return (
      <div className={styles['pho-content-wrapper']}>
        <Topbar viewName='albums'>
          <AlbumsToolbar />
        </Topbar>
        <Content list={this.props.albums} sharedByMe={this.state.sharedByMe} sharedWithMe={this.state.sharedWithMe} />
      </div>
    )
  }
}

const mapDocumentsToProps = (ownProps) => ({ albums: fetchAlbums() })

export default cozyConnect(mapDocumentsToProps)(AlbumsView)
