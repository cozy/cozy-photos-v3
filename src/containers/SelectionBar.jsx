import styles from '../styles/selectionbar'

import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import classNames from 'classnames'
import { translate } from 'cozy-ui/react/I18n'

import Alerter from '../components/Alerter'
import { hideSelectionBar } from '../actions/selection'

import { getAlbum, openAddToAlbum, removeFromAlbum } from '../ducks/albums'

const SelectionBar = ({ t, selected, selectedCount, album, onHide, onAddToAlbum, onRemoveFromAlbum, router }) => (
  <div
    className={classNames(styles['coz-selectionbar'], {
      [styles['coz-selectionbar--active']]: selectedCount !== 0
    })}
    role='toolbar'
  >
    <span className={styles['coz-selectionbar-count']}>
      {t('SelectionBar.selected_count', { smart_count: selectedCount })}
    </span>
    <span className={styles['coz-selectionbar-separator']} />
    <button
      disabled={selectedCount === 0}
      className={styles['coz-action-album-add']}
      onClick={() => onAddToAlbum(selected)}
    >
      {t('SelectionBar.add_to_album')}
    </button>
    {router.location.pathname.startsWith('/albums') &&
    <button
      disabled={selectedCount === 0}
      className={styles['coz-action-album-remove']}
      onClick={() => onRemoveFromAlbum(selected, album)}
    >
      {t('SelectionBar.remove_from_album')}
    </button>
    }
    <button className={styles['coz-action-close']} onClick={onHide}>
      {t('SelectionBar.close')}
    </button>
  </div>
)

const mapStateToProps = (state, ownProps) => ({
  selected: state.ui.selected,
  selectedCount: state.ui.selected.length,
  album: ownProps.params.albumId ? getAlbum(state, ownProps.params.albumId) : null
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  onHide: () => dispatch(hideSelectionBar()),
  onAddToAlbum: selected => dispatch(openAddToAlbum(selected)),
  onRemoveFromAlbum: (selected, album) =>
  dispatch(removeFromAlbum(album, selected))
   .then(() => Alerter.success('Albums.remove_photos.success', { album_name: album.name }))
   .catch(() => Alerter.error('Albums.remove_photos.error.generic'))
})

export default withRouter(connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(SelectionBar)))
