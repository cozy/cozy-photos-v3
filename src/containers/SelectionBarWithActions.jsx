import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { translate } from 'cozy-ui/react/I18n'

import { openAddToAlbum, removeFromAlbum } from '../ducks/albums'
import { SelectionBar, hideSelectionBar, downloadSelection } from '../ducks/selection'
import { DeleteConfirm, deletePhotos } from '../ducks/timeline'
import Alerter from '../components/Alerter'
import confirm from '../lib/confirm'

const mapDispatchToProps = (dispatch, ownProps) => ({
  actions: {
    'album-add': {
      action: selected => dispatch(openAddToAlbum(selected))
    },
    'download': {
      action: selected => downloadSelection(selected)
    },
    'delete': {
      action: selected => confirm(<DeleteConfirm t={ownProps.t} count={selected.length} related={ownProps.related} />,
        () => dispatch(deletePhotos(selected))
      ),
      displayCondition: () => ownProps.router.location.pathname.startsWith('/photos')
    },
    'album-remove': {
      action: selected => dispatch(removeFromAlbum(ownProps.album, selected))
        .then(() => {
          Alerter.success('Albums.remove_photos.success', { album_name: ownProps.album.name })
          // TODO: we should find a better way, perhaps closing the bar by default on action success,
          // and provide an optional 'cancelSelectionOnSuccess' option for the bar actions
          dispatch(hideSelectionBar())
        })
        .catch(() => Alerter.error('Albums.remove_photos.error.generic')),
      displayCondition: () => (ownProps.album !== undefined && ownProps.readOnly === false)
    }
  }
})

export default translate()(withRouter(connect(null, mapDispatchToProps)(SelectionBar)))
