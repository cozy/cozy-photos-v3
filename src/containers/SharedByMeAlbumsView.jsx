import React, { Component } from 'react'
import { connect } from 'react-redux'
import styles from '../styles/layout'

import { AlbumsToolbar, fetchAlbums, getAlbumsList } from '../ducks/albums'
import { filterSharedByMeDocuments, filterSharedWithMeDocuments } from '../ducks/sharing'

import AlbumsList from '../components/AlbumsList'
import Loading from '../components/Loading'
import ErrorComponent from '../components/ErrorComponent'
import Topbar from '../components/Topbar'

const Content = ({ list, sharedByMe, sharedWithMe }) => {
  const { fetchStatus, entries } = list
  switch (fetchStatus) {
    case 'pending':
    case 'loading':
      return <Loading loadingType='albums_fetching' />
    case 'failed':
      return <ErrorComponent errorType='albums' />
    default:
      return <AlbumsList albums={entries} sharedByMe={sharedByMe} sharedWithMe={sharedWithMe} />
  }
}

export class AlbumsView extends Component {
  constructor (props) {
    super(props)
    this.state = {
      sharedByMe: [],
      sharedWithMe: []
    }
  }

  componentWillMount () {
    this.props.fetchAlbums()
  }

  componentWillReceiveProps (newProps) {
    // TODO: this is not the cleanest way of doing this...
    if (newProps.albums && newProps.albums.entries !== 0) {
      Promise.all([
        filterSharedByMeDocuments(newProps.albums.entries.map(a => a.id), 'io.cozy.photos.albums'),
        filterSharedWithMeDocuments(newProps.albums.entries.map(a => a.id), 'io.cozy.photos.albums')
      ])
        .then(sharedIds => this.setState({ sharedByMe: sharedIds[0], sharedWithMe: sharedIds[1] }))
    }
  }

  render () {
    if (this.props.children) return this.props.children
    if (!this.props.albums) {
      return null
    }
    console.log(this.state.sharedByMe)
    return (
      <div className={styles['pho-content-wrapper']}>
        <Topbar viewName='sharedbyme'>
          <AlbumsToolbar />
        </Topbar>
        <Content list={this.props.albums} sharedByMe={this.state.sharedByMe} sharedWithMe={this.state.sharedWithMe} />
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  albums: getAlbumsList(state)
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchAlbums: () => dispatch(fetchAlbums())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AlbumsView)
