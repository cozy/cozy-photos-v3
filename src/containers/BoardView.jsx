import React, { Component } from 'react'
import { connect } from 'react-redux'

import {
  isSelectionBarVisible,
  getSelectedIds,
  toggleItemSelection,
  addToSelection,
  removeFromSelection
} from '../ducks/selection'
import { isRelated } from '../ducks/timeline'

import SelectionBarWithActions from './SelectionBarWithActions'
import PhotoBoard from '../components/PhotoBoard'
import AddToAlbumModal from '../containers/AddToAlbumModal'

class BoardView extends Component {
  render () {
    const {
      album,
      related,
      readOnly,
      selected,
      isAddToAlbumModalOpened,
      onPhotoToggle,
      onPhotosSelect,
      onPhotosUnselect,
      selectionModeActive,
      photosContext
    } = this.props

    const {
      photoLists,
      photos
    } = this.props

    return (
      <div role='contentinfo'>
        {isAddToAlbumModalOpened && <AddToAlbumModal />}
        {selectionModeActive && <SelectionBarWithActions album={album} related={related} readOnly={readOnly} />}
        <PhotoBoard
          lists={photoLists}
          selected={selected}
          photosContext={photosContext}
          showSelection={selectionModeActive}
          onPhotoToggle={onPhotoToggle}
          onPhotosSelect={onPhotosSelect}
          onPhotosUnselect={onPhotosUnselect}
          fetchStatus={photos.fetchStatus}
          hasMore={photos.hasMore}
          onFetchMore={photos.fetchMore}
        />
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  selected: getSelectedIds(state),
  related: isRelated(state, ownProps.photos.data),
  selectionModeActive: isSelectionBarVisible(state),
  isAddToAlbumModalOpened: state.ui.isAddToAlbumModalOpened
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  onPhotoToggle: (id, selected) => {
    dispatch(toggleItemSelection(id, selected))
  },
  onPhotosSelect: (ids) => {
    dispatch(addToSelection(ids))
  },
  onPhotosUnselect: (ids) => {
    dispatch(removeFromSelection(ids))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BoardView)
