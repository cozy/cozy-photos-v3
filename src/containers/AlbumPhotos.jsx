import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import styles from '../styles/layout'

import { AlbumToolbar, getAlbum, getAlbumPhotos, fetchAlbums, fetchAlbumPhotos, updateAlbum } from '../ducks/albums'

import BoardView from './BoardView'
import Topbar from '../components/Topbar'
import Alerter from '../components/Alerter'

export class AlbumPhotos extends Component {
  constructor (props) {
    super(props)
    this.state = {
      editing: false
    }
  }

  componentWillMount () {
    if (!this.props.album) {
      this.props.fetchAlbums()
        .then(() => this.props.fetchPhotos(this.props.router.params.albumId))
    } else {
      this.props.fetchPhotos(this.props.router.params.albumId)
    }
  }

  editAlbumName () {
    this.setState({editing: true})
  }

  renameAlbum (name) {
    if (name.trim() === '') {
      Alerter.error('Error.album_rename_abort')
      return
    } else if (name === this.props.album.name) {
      this.setState({editing: false})
      return
    }

    let updatedAlbum = { ...this.props.album, name: name }
    this.props.updateAlbum(updatedAlbum)
      .then(() => {
        this.setState({editing: false})
      })
      .catch(() => {
        Alerter.error('Error.generic')
      })
  }

  render () {
    if (!this.props.album) {
      return null
    }
    const { album, photos, fetchPhotos } = this.props
    const { editing } = this.state
    return (
      <div className={styles['pho-content-wrapper']}>
        {album.name &&
          <Topbar viewName='albumContent' albumName={album.name} editing={editing} onEdit={this.renameAlbum.bind(this)} >
            <AlbumToolbar album={album} onRename={this.editAlbumName.bind(this)} />
          </Topbar>
        }
        {photos &&
          <BoardView
            photoLists={[{ photos: photos.entries }]}
            fetchStatus={photos.fetchStatus}
            hasMore={photos.hasMore}
            photosContext='album'
            onFetchMore={() => fetchPhotos(album._id, photos.entries.length)}
          />
        }
        {this.renderViewer(this.props.children)}
      </div>
    )
  }

  renderViewer (children) {
    if (!children) return null
    return React.Children.map(children, child => React.cloneElement(child, {
      photos: this.props.photos.entries
    }))
  }
}

const mapStateToProps = (state, ownProps) => ({
  album: getAlbum(state, ownProps.router.params.albumId),
  photos: getAlbumPhotos(state, ownProps.router.params.albumId)
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchAlbums: () => dispatch(fetchAlbums()),
  fetchPhotos: (albumId, skip) => dispatch(fetchAlbumPhotos(albumId, skip)),
  updateAlbum: (album) => dispatch(updateAlbum(album))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AlbumPhotos))
