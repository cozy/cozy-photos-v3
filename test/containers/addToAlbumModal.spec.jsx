'use strict'

/* eslint-env jest */
// Specifying mock directly, manual mocks in __mocks__ subfolder, as described
// in https://facebook.github.io/jest/docs/manual-mocks.html seems to not
// work properly.
jest.mock('cozy-ui/react/Modal')

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from '../lib/I18n'
import { AddToAlbumModal } from '../../src/containers/AddToAlbumModal'

const mockFetchedPhotos = [
  {
    _id: '33dda00f0eec15bc3b3c59a615001ac8',
    created_at: '0001-01-01T00:00:00Z',
    name: 'MonImage.jpg',
    size: '150000',
    updated_at: '0001-01-01T00:00:00Z'
  }
]

describe('AddToAlbumModal component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <AddToAlbumModal t={mockT} photos={mockFetchedPhotos} onDismiss={jest.fn()} onSubmitNewAlbum={jest.fn()} />
    )
    expect(component.node).toMatchSnapshot()
  })
})
