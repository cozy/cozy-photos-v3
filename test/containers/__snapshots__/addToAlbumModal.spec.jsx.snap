// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToAlbumModal component should be rendered correctly 1`] = `
<Unknown
  secondaryAction={[Function]}
  title="Add to album"
>
  <div
    className="coz-modal-section coz-create-album"
  >
    <_translate
      onSubmitNewAlbum={[Function]}
    />
  </div>
</Unknown>
`;
